### PR TITLE
Add an environment variable to set the context window size

### PR DESCRIPTION
**Pull request by ghost**
## Context

We're getting errors because [the context window size is being exceeded](https://technologyprogramme.atlassian.net/jira/software/projects/REDBOX/boards/409?selectedIssue=REDBOX-417).

## Changes proposed in this pull request

* Adds an environment variable to declare the maximum
* Adds a tiktoken dependency to count tokens
* Uses it in the chat history, and nowhere else for now

## Guidance to review

* Check you're happy with how I've implemented this
* Run the django app and check it works

## Things to check

- [x] I have added any new ENV vars in all deployed environments
- [x] I have tested any code added or changed
- [x] [I have run integration tests](https://github.com/i-dot-ai/redbox-copilot/actions/runs/9683170696)

### CODE REVIEW

**Comment by brunns** [comment]
Do we need to add this to any of the `.env` files?

**Comment by brunns** [comment]
If it picks up sensible defaults elsewhere, that's fine I think. But if we need them set, we should set them in `.env.integration` at least.

Thinking about it, if the tests are passing, we clearly don't need them explicitly set in `.env` so it's fine, right?

File: core_api/src/runnables.py [hunk 1]
```diff
@@ -10,24 +10,47 @@
...
-def make_chat_prompt_from_messages_runnable(system_prompt: str, question_prompt: str):
+def make_chat_prompt_from_messages_runnable(
+    system_prompt: str,
+    question_prompt: str,
+    context_window_size: int,
+    tokeniser: Encoding,
+):
     system_prompt_message = [("system", system_prompt)]
+    prompts_budget = len(tokeniser.encode(system_prompt)) - len(tokeniser.encode(question_prompt))
+    token_budget = context_window_size - prompts_budget
```
Comment (jamesrichards4) [inline core_api/src/runnables.py:29]: This needs to take into account the output max tokens (llm.max_tokens)

File: core_api/src/runnables.py [hunk 2]
```diff
@@ -10,24 +10,47 @@
...
     def chat_prompt_from_messages(input_dict: dict):
         """
         Create a ChatPrompTemplate as part of a chain using 'chat_history'.
         Returns the PromptValue using values in the input_dict
         """
+        context_budget = token_budget - len(tokeniser.encode(input_dict["question"]))
+
+        if context_budget <= 0:
+            message = "Question length exceeds context window."
+            raise AIError(message)
```
Comment (jamesrichards4) [inline core_api/src/runnables.py:41]: This gives a 500 in the API response? What about Django?

Comment (brunns) [reply core_api/src/runnables.py:41]: Sorry, not sure I understand the Django question.

**Review by jamesrichards4 — APPROVED** [review]

**Pull request merged**